src/wallet/orchard.h

```diff
@@ -65,21 +65,42 @@ class OrchardWallet
     OrchardWallet(const OrchardWallet&) = delete;
     OrchardWallet& operator=(const OrchardWallet&) = delete;
 
-    void CheckpointNoteCommitmentTree() {
-        orchard_wallet_checkpoint(inner.get());
+    /**
+     * Checkpoint the note commitment tree. This returns `false` and leaves the
+     * note commitment tree unmodified if the block height does not match the
+     * last block height scanned for transactions. This must be called exactly
+     * once per block.
+     */
+    bool CheckpointNoteCommitmentTree(int nBlockHeight) {
+        return orchard_wallet_checkpoint(inner.get(), (uint32_t) nBlockHeight);
+    }
+
+    /**
+     * Rewind to the most recent checkpoint, and mark as unspent any notes
+     * previously identified as having been spent by transactions in the
+     * latest block.
+     */
+    bool IsCheckpointed() const {
+        return orchard_wallet_is_checkpointed(inner.get());
     }
 
-    bool RewindToLastCheckpoint() {
-        return orchard_wallet_rewind(inner.get());
+    /**
+     * Rewind to the most recent checkpoint, and mark as unspent any notes
+     * previously identified as having been spent by transactions in the
+     * latest block.
+     */
+    bool Rewind(int nBlockHeight) {
+        return orchard_wallet_rewind(inner.get(), (uint32_t) nBlockHeight);
     }
 
     /**
      * Add notes that are decryptable with IVKs for which the wallet
      * contains the full viewing key to the wallet.
      */
-    void AddNotesIfInvolvingMe(const CTransaction& tx) {
+    void AddNotesIfInvolvingMe(const int nBlockHeight, const CTransaction& tx) {
         orchard_wallet_add_notes_from_bundle(
                 inner.get(),
+                (uint32_t) nBlockHeight,
```

Comment:
```suggestion
    void AddNotesIfInvolvingMe(const CTransaction& tx, const std::optional<int> nBlockHeight = std::nullopt) {
        orchard_wallet_add_notes_from_bundle(
                inner.get(),
                nBlockHeight.has_value() ? &((uint32_t) nBlockHeight.value()) : nullptr,
```
(IDK if the inline reference to the memory of an anonymous variable works; might need to define a temporary.)